x/wasm/IBC.md

```diff
@@ -0,0 +1,425 @@
+   the contract can reject the connection or accept it and return the protocol version it will communicate with.
+   Implementing this requires that we support [ADR 025](https://github.com/cosmos/cosmos-sdk/blob/master/docs/architecture/adr-025-ibc-passive-channels.md).
+   TODO: check with Agoric and cwgoes to upstream any Cosmos SDK changes so this can work out of the box.
+3. `OnChanOpened` is called on the contract for both `OnChanOpenAck` and `OnChanOpenConfirm` containing
+   the final version string (counterparty version). This gives a chance to abort the process
+   if we realize this doesn't work. Or save the info (we may need to define this channel uses an older version
+   of the protocol for example).
+4. `OnChanClosed` is called on both sides if the channel is closed for any reason, allowing them to
+   perform any cleanup. This will be followed by `IBCPacketTimeout` callbacks for all the in-progress 
+   packets that were not processed before it closed, as well as all pending acks (pending the relayer
+   to provide that).
+
+We require the following callbacks on the contract
+
+* `OnChanNegotiate` - called on receiving end for `OnChanOpenTry`. 
+* `OnChanOpened` - called on both sides after the initial 2 steps have passed to confirm the version used
+* `OnChanClosed` - this is called when an existing channel is closed for any reason
+
+Note @cwgoes I am rather confused by 
+[the handshake docs](https://docs.cosmos.network/master/ibc/custom.html#channel-handshake-version-negotiation)
+In particular "ChanOpenTry callback should verify that the MsgChanOpenTry.Version is valid and that 
+MsgChanOpenTry.CounterpartyVersion is valid.". Where does the CounterpartyVersion come from?
+I thought the module would accept the Version and decide on it's own CounterpartyVersion. We assume the
+relayer makes that decision????
+
+### Queries
+
+We may want to expose some basic queries of IBC State to the contract.
+We should check if this is useful to the contract and if it opens up
+any possible DoS:
+
+* `GetPortID` - return PortID given a contract address
+* `ListChannels` - return a list of all (portID, channelID) pairs
+  that are bound to a given port.
+* `ListPendingPackets` - given a (portID, channelID) identifier, return all packets
+  in that channel, that have been sent by this chain, but for which no acknowledgement 
+  or timeout has yet been received
+
+## Contract Details
+
+Here we map out the workflow with the exact arguments passed with those calls 
+from the Go side (which we will use with our mock), and then a 
+proposal for multiplexing this over fewer wasm exports (define some rust types)
+
+Messages:
+
+```go
+package messages
+
+type IBCSendMsg struct {
+    // This is our contract-local ID
+    ChannelID string
+    Msg []byte
+    // optional fields (or do we need exactly/at least one of these?)
+    TimeoutHeight uint64
+    TimeoutTimestamp uint64
```

Comment:
We need the `Sequence` here as well to have all fields for an [ibc packet](https://github.com/cosmos/cosmos-sdk/blob/master/x/ibc/04-channel/types/packet.go#L30)
`SourcePortID` is known to us from the contract context and we can lookup the counterpart with the channelID:portID pair

Reply:
The sequence should not be set by the contract, but auto-assigned by the keeper/ibc module. 

>  SourcePortID is known to us from the contract context and we can lookup the counterpart with the channelID:portID pair

Yes, I should add that as a document. The ChannelID (and context-contained PortID) should be enough to uniquely identify the channel and remote counterparty.

Reply:
👍  correct. we can use `channelKeeper.GetNextSequenceSend(ctx, sourcePort, sourceChannel)` to get the next sequence to build the package. Do I understand the workflow correct?
1. the contract returns an IBCSendMsg
2. in our handler_plugin we create the `channeltypes.Packet`
3. we send the packet to `channelKeeper.SendPacket`
3a. within the handler_plugin
3b return it and send it from the keeper
4. relayer sends packet to other chain....

Reply:
That sounds correct.

Now that you mention the details of the HandlerPlugin interface, it doesn't by itself have the ability to query into the channelKeeper to get that info. I don't want to push that responsibility down to the contract and I guess we will have to extend the HandlerPlugin interface to have more capabilities during the transform.